### DISCUSSION TITLE
Unidirectional websocket connections where only the server pushes data to the clients

### DISCUSSION THREAD

**Discussion by JohnnyRacer** [fastapi/fastapi#6127] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

### Description

Hello,
Is there a way I could send data to clients over websocket without listening for when clients send data back. I'm trying to have a websocket endpoint where the server is pushing data to the client in a unidirectional way without the option for the client to send responses back. There doesn't seem to be any code that I could find that supports this since all the documentation seems to require that the server is listening for a `websocket.recieve_text()`. Any help would be much appreciated, thanks.

### Operating System

Linux

### FastAPI Version

0.81.0

### Python Version

3.8.13

**Comment by JarroVGIT** [comment]
I don’t understand, why not simply omit `websocket.recieve_text()` in your example?

**Comment by JohnnyRacer** [comment]
@JarroVGIT 
I have tried that but the websocket connection doesn't seem to open correctly.  If I have a snippet of code like below where the time is being suppose to be sent over the connection every 10s, it ends up being sent instantaneously and spams the client.  
```python
@router.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        if int(time.time()) % 10 == 0:
            print('Sending the time!')
            await websocket.send_text(str(time.time()))
```
But if I include `websocket.recieve_text()`, I get the expected correct response of the time on every message sent using the snippet below.

```python
@router.websocket("/wsx")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(str(time.time()))
```

I would like a way to send messages correctly without `websocket.recieve_text()` if possible. I can try to clarify it up with more examples if needed. Thanks.

PS: I accidentally closed the issue, it is still not behaving as expected. :(

**Comment by JarroVGIT** [comment, marked as answer]
Your logic is flawed I am afraid (the modulo operation). The following works fine (and runs as-is):

```python
import asyncio
import datetime

from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from starlette.websockets import WebSocketClose
from websockets.exceptions import ConnectionClosedError

app = FastAPI()


@app.get("/")
async def root():
    return {"hello": "world"}


@app.websocket("/wsx")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        t = datetime.datetime.utcnow().time()
        try:
            if t.second % 5 == 0:
                print("Sending the time!")
                await websocket.send_text(str(t))
                await asyncio.sleep(1)
        except ConnectionClosedError:
            print("Client disconnected.")
            break


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```
Left is my client connecting with the websocket endpoint. Right is the logs of the above snippet. I disconnected using ctrl-c (otherwise it would run forever).
<img width="1086" alt="image" src="https://user-images.githubusercontent.com/13659033/192052180-b5fe6bc1-4b9b-4cb7-bb3a-b45d724795c7.png">

**Comment by JohnnyRacer** [comment]
Thank you so much! The code snippet you provided works as expected. I found that the behavior reverts back to how it was if I remove `asyncio.sleep(1)` (client gets spammed instead of being sent in regular intervals) , is there a reason why this line is required for the endpoint to send the data properly? Since I believe it is causing a minor issue where the endpoint would occasionally send the data at the wrong interval (1 second instead of 5 seconds). I've attached a screenshot below showing this behavior when I use a browser to access the websocket connection.
![image](https://user-images.githubusercontent.com/77214388/192059576-1d6f1f2f-7421-46bd-a13d-75ed0500d70f.png)

**Comment by JarroVGIT** [comment]
Yes, but is has nothing to do with websockets. 

The issue is that the while loop is ran multiple times during a second. That’s why I sleep for 1 second, so I know the time is at least in the next second when the while loop is run again. 

This is not perfect logic either, if you require data to be send exactly every second, then you would have to design for that. But that was not the issue in this topic, that is more of ansubject to be asked on SO.

**Comment by JohnnyRacer** [comment]
Thanks for clarifying, that already answered so much of my questions.

**Comment by tiangolo** [comment]
Thanks for the help here @JarroVGIT  ! 👏  🙇 

Thanks for reporting back and closing the issue @JohnnyRacer  👍

**Comment by Viicos** [comment]
Related to this, I've also noticed a strange behavior when omitting `await websocket.receive_*()`. Without it, the `while True` loop would freeze any other stuff running (that is, no other client could connect to the websocket endpoint), and stuff running in the background with [`repeat_every`](https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/#the-repeat_every-decorator) would freeze as well.
